File: go/span.go
package tree2

import "fmt"

// Span marks a region of a source resource.
//
// Row is 1-based. Col is a 1-based offset inside that row and Length is a
// region size, both measured in Unicode code points — the reference TypeScript
// implementation measures them in UTF-16 code units, which is a JavaScript
// artifact (see SPEC.md §2.3).
//
// A Span is a value: copying it is cheap and copies never alias.
type Span struct {
	// URI names the source resource.
	URI string
	// Source is the full text of that resource.
	Source string
	// Row is the 1-based line number.
	Row int
	// Col is the 1-based column, in code points.
	Col int
	// Length is the size of the marked region, in code points.
	Length int
}

// SpanUnknown marks the beginning of an unknown resource. It is the span of
// nodes built without any source behind them.
var SpanUnknown = SpanBegin("?", "")

// NewSpan makes a span over an explicit region.
func NewSpan(uri, source string, row, col, length int) Span {
	return Span{URI: uri, Source: source, Row: row, Col: col, Length: length}
}

// SpanBegin makes an empty span at the beginning of a resource.
func SpanBegin(uri, source string) Span {
	return Span{URI: uri, Source: source, Row: 1, Col: 1, Length: 0}
}

// SpanEnd makes an empty span just past the end of a resource.
func SpanEnd(uri, source string) Span {
	return Span{URI: uri, Source: source, Row: 1, Col: runeLen(source) + 1, Length: 0}
}

// SpanEntire makes a span covering a whole resource.
func SpanEntire(uri, source string) Span {
	return Span{URI: uri, Source: source, Row: 1, Col: 1, Length: runeLen(source)}
}

// Span makes another span over the same resource.
func (s Span) Span(row, col, length int) Span {
	return Span{URI: s.URI, Source: s.Source, Row: row, Col: col, Length: length}
}

// After makes a span of the given length starting right after this one.
func (s Span) After(length int) Span {
	return s.Span(s.Row, s.Col+s.Length, length)
}

// Slice makes a sub-span of this one. Negative bounds count from the end, so
// Slice(0, -1) drops the last code point — the same default the reference uses.
// The bounds must stay inside the span and end must not precede begin.
func (s Span) Slice(begin, end int) (Span, error) {
	if begin < 0 {
		begin += s.Length
	}
	if end < 0 {
		end += s.Length
	}

	if begin < 0 || begin > s.Length {
		return Span{}, s.Error(fmt.Sprintf("Begin value '%d' out of range", begin))
	}
	if end < 0 || end > s.Length {
		return Span{}, s.Error(fmt.Sprintf("End value '%d' out of range", end))
	}
	if end < begin {
		return Span{}, s.Error(fmt.Sprintf("End value '%d' can't be less than begin value", end))
	}

	return s.Span(s.Row, s.Col+begin, end-begin), nil
}

// Error makes an error pointing at this span.
func (s Span) Error(message string) error {
	return &SpanError{Message: message, Span: s}
}

// String renders the span as `uri#row:col/length`.
func (s Span) String() string {
	return fmt.Sprintf("%s#%d:%d/%d", s.URI, s.Row, s.Col, s.Length)
}

// runeLen counts code points, which is what columns and lengths are measured
// in. Continuation bytes are the ones that do not start a code point.
func runeLen(str string) int {
	count := 0
	for i := 0; i < len(str); i++ {
		if str[i]&0xC0 != 0x80 {
			count++
		}
	}
	return count
}

// SpanError is an error carrying the position it happened at.
type SpanError struct {
	Message string
	Span    Span
}

func (e *SpanError) Error() string {
	return fmt.Sprintf("%s (%s)", e.Message, e.Span)
}

File: rust/src/span.rs
//! Positions inside a source resource.

use std::fmt;
use std::sync::Arc;

use crate::error::Error;

/// A region of a source resource: a row, a column and a length.
///
/// Columns and lengths are measured in **Unicode scalar values** (code points),
/// not in UTF-16 code units as the JavaScript reference does. See the
/// specification, §2.3.
///
/// A span carries the whole source text, shared between every span of the same
/// resource, so cloning one is two atomic increments.
///
/// ```
/// let span = tree2::Span::entire("test", "hello\n");
/// assert_eq!(span.to_string(), "test#1:1/6");
/// ```
#[derive(Clone, PartialEq, Eq)]
pub struct Span {
    uri: Arc<str>,
    source: Arc<str>,
    row: usize,
    col: usize,
    length: usize,
}

impl Span {
    /// Makes a span over an explicit region.
    ///
    /// `row` and `col` are 1-based; `length` is a count of code points.
    pub fn new(
        uri: impl Into<Arc<str>>,
        source: impl Into<Arc<str>>,
        row: usize,
        col: usize,
        length: usize,
    ) -> Self {
        Self {
            uri: uri.into(),
            source: source.into(),
            row,
            col,
            length,
        }
    }

    /// Span for the beginning of a resource.
    pub fn begin(uri: impl Into<Arc<str>>, source: impl Into<Arc<str>>) -> Self {
        Self::new(uri, source, 1, 1, 0)
    }

    /// Span for the end of a resource.
    pub fn end(uri: impl Into<Arc<str>>, source: impl Into<Arc<str>>) -> Self {
        let source = source.into();
        let col = source.chars().count() + 1;
        Self::new(uri, source, 1, col, 0)
    }

    /// Span covering an entire resource.
    pub fn entire(uri: impl Into<Arc<str>>, source: impl Into<Arc<str>>) -> Self {
        let source = source.into();
        let length = source.chars().count();
        Self::new(uri, source, 1, 1, length)
    }

    /// Span for the beginning of an unknown resource, named `?`.
    pub fn unknown() -> Self {
        Self::begin("?", "")
    }

    /// Name of the source resource.
    pub fn uri(&self) -> &str {
        &self.uri
    }

    /// Full text of the source resource.
    pub fn source(&self) -> &str {
        &self.source
    }

    /// 1-based line number.
    pub fn row(&self) -> usize {
        self.row
    }

    /// 1-based column, in code points.
    pub fn col(&self) -> usize {
        self.col
    }

    /// Length of the marked region, in code points.
    pub fn length(&self) -> usize {
        self.length
    }

    /// Makes another span over the same resource.
    pub fn span(&self, row: usize, col: usize, length: usize) -> Self {
        Self {
            uri: self.uri.clone(),
            source: self.source.clone(),
            row,
            col,
            length,
        }
    }

    /// Makes a span starting right after this one, on the same row.
    pub fn after(&self, length: usize) -> Self {
        self.span(self.row, self.col.saturating_add(self.length), length)
    }

    /// Makes a span over a part of this one.
    ///
    /// Negative indices count from the end of this span. Note that the
    /// reference's default `end` is `-1`, which drops the last code point —
    /// see [`Span::slice_from`].
    ///
    /// # Errors
    ///
    /// Fails when an index falls outside this span, or when `end` precedes
    /// `begin`.
    pub fn slice(&self, begin: isize, end: isize) -> Result<Self, Error> {
        let len = self.length as isize;

        let begin = if begin < 0 {
            begin.saturating_add(len)
        } else {
            begin
        };
        let end = if end < 0 {
            end.saturating_add(len)
        } else {
            end
        };

        if begin < 0 || begin > len {
            return Err(self.error(format_args!("Begin value '{begin}' out of range")));
        }
        if end < 0 || end > len {
            return Err(self.error(format_args!("End value '{end}' out of range")));
        }
        if end < begin {
            return Err(self.error(format_args!(
                "End value '{end}' can't be less than begin value"
            )));
        }

        Ok(self.span(
            self.row,
            self.col.saturating_add(begin as usize),
            (end - begin) as usize,
        ))
    }

    /// [`Span::slice`] with the reference's default end of `-1`.
    ///
    /// # Errors
    ///
    /// As [`Span::slice`].
    pub fn slice_from(&self, begin: isize) -> Result<Self, Error> {
        self.slice(begin, -1)
    }

    /// Makes an error over this span, rendered as `{message} ({span})`.
    pub fn error(&self, message: impl fmt::Display) -> Error {
        Error::new(format!("{message} ({self})"), self.clone())
    }
}

impl fmt::Display for Span {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}#{}:{}/{}", self.uri, self.row, self.col, self.length)
    }
}

impl fmt::Debug for Span {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Span({self})")
    }
}
